pkg/compressor/compressor.go

```diff
@@ -0,0 +1,138 @@
+// and write the compressed data into one end of pipe.
+func CompressSnapshot(data io.ReadCloser, compressionPolicy string) io.ReadCloser {
+	pReader, pWriter := io.Pipe()
+
+	var gWriter io.WriteCloser
+	logger := logrus.New().WithField("actor", "compressor")
+	logger.Infof("start compressing the snapshot using %v Compression Policy", compressionPolicy)
+
+	switch compressionPolicy {
+	case GzipCompressionPolicy:
+		gWriter = gzip.NewWriter(pWriter)
+
+	case LzwCompressionPolicy:
+		gWriter = lzw.NewWriter(pWriter, lzw.LSB, LzwLiteralWidth)
+
+	case ZlibCompressionPolicy:
+		gWriter = zlib.NewWriter(pWriter)
+
+	}
+
+	go func() {
+		var err error
+		defer pWriter.CloseWithError(err)
+		defer gWriter.Close()
+		defer data.Close()
+		_, err = io.Copy(gWriter, data)
+		if err != nil {
+			logger.Infof("compression failed: %v", err)
+			return
+		}
+	}()
+
+	return pReader
+}
+
+// DecompressSnapshot take compressed data and compressionPolicy as input and
+// it decompresses the data according to compression Policy and return uncompressed data.
+func DecompressSnapshot(data io.ReadCloser, compressionPolicy string) (io.ReadCloser, error) {
+	var deCompressedData io.ReadCloser
+	var err error
+
+	logger := logrus.New().WithField("actor", "de-compressor")
+	logger.Infof("start decompressing the snapshot with %v compressionPolicy", compressionPolicy)
+
+	switch compressionPolicy {
+	case ZlibCompressionPolicy:
+		deCompressedData, err = zlib.NewReader(data)
+		if err != nil {
+			logger.Infof("Unable to decompress: %v", err)
+			return data, err
+		}
+
+	case GzipCompressionPolicy:
+		deCompressedData, err = gzip.NewReader(data)
+		if err != nil {
+			logger.Infof("Unable to decompress: %v", err)
+			return data, err
+		}
+
+	case LzwCompressionPolicy:
+		deCompressedData = lzw.NewReader(data, lzw.LSB, LzwLiteralWidth)
+
+	// It is actually unreachable but just to be on safe side:
+	// for unsupported CompressionPolicy return the same data
+	default:
+		return data, fmt.Errorf("Unsupported Compression Policy")
+	}
+
+	return deCompressedData, nil
+}
+
+// GetCompressionSuffix returns the suffix for snapshot w.r.t Compression Policy
+// if compression is not enabled, it will simply return empty string.
+func GetCompressionSuffix(compressionEnabled bool, compressionPolicy string) string {
+	emptyString := ""
+	if !compressionEnabled {
+		return emptyString
+	}
+
+	switch compressionPolicy {
+	case ZlibCompressionPolicy:
+		return ZlibCompressionExtension
+
+	case LzwCompressionPolicy:
+		return LzwCompressionExtension
+
+	case GzipCompressionPolicy:
+		return GzipCompressionExtension
+	}
+
+	return emptyString
+}
+
+// IsSnapshotCompressed is helpful in determining whether the snapshot is compressed or not.
+// it will return boolean and compressionPolicy corresponding to compressionSuffix.
+func IsSnapshotCompressed(compressionSuffix string) (bool, string) {
```

Comment:
Should we also return error for the default case (unknown suffix/policy)?

Reply:
actually unknown compression Policy get handled [here](https://github.com/gardener/etcd-backup-restore/pull/293/files#diff-7adcf3334be4957b8c200cd4ad81c23a84d4be71d629640291a51ba3b20a198eR38-R51) .... and for safety purpose: we are returning `false and emptyString` .... so I think for unknown compression Policy it will treat them as uncompressed and will return `false and emptyString`

Reply:
Thanks for addressing the default case above. Can you also address it here? The same reason applies.